lib/sigh/resign.rb

```diff
@@ -112,19 +120,63 @@ def ask_for_signing_identity
       ask('Signing Identity: ')
     end
 
-    # Array of available signing identities
-    def installed_identies
-      available = `security find-identity -v -p codesigning`
-      ids = []
+    # Hash of available signing identities
+    def installed_identities
+      available = `security find-identity -p codesigning`
+      ids = {}
       available.split("\n").each do |current|
         begin
-          (ids << current.match(/.*\"(.*)\"/)[1])
+          match = current.match(/.*([0-9A-Z]{40}) \"(.*)\"(\s+\((.*)\))?/)
+          sha1 = match[1]
+          name = match[2]
+          issue = match[4]
+          ids[sha1] = SigningIdentity.new(sha1, name, issue)
         rescue
           nil
         end # the last line does not match
       end
 
-      ids
+      ids.values.select(&:may_be_valid?).sort_by do |identity|
+        [identity.name, identity.valid? ? 0 : 1]
+      end
+    end
+
+    def installed_identity_descriptions
+      descriptions = []
+      installed_identities.group_by(&:name).each do |name, identities|
+        descriptions << (identities.any?(&:valid?) ? name.black : name.white)
+        # Show SHA-1 for homonymous identities
+        descriptions += identities.map do |identity|
+          text = "\t#{identity.sha1}"
+          identity.valid? ? text.black : text.white
+        end if identities.count > 1
+      end
+      descriptions
+    end
+
+    class SigningIdentity
+      attr_reader :sha1, :name, :issue
+
+      def initialize(sha1, name, issue)
+        @sha1 = sha1
+        @name = name
+        @issue = issue
+      end
+
+      def valid?
+        @issue.nil?
+      end
+
+      def may_be_valid?
+        # Because of WWDR Intermediate Certificate expiration, valid certificates may be reported as expired.
+        # So we need to accept expired certificates for the moment.
+        # https://developer.apple.com/support/certificates/expiration
+        [nil, 'CSSMERR_TP_CERT_EXPIRED'].include? @issue
```

Comment:
Do you have any documentation on the `CSSMERR_TP_CERT_EXPIRED` value? Did you discover it via terminal output?